### DISCUSSION TITLE
Getting i3-like UX for navigating groups

### DISCUSSION THREAD

**Discussion by nadiamoe** [hyprwm/Hyprland#2517] [Q&A]
Hi there!

Long time i3 user here dipping their toes in wayland and hyprland :)

I've played with hyprland for a couple hours already, but there's something I'm not being able to figure out regarding how group navigation works. I've set up some very basic bindings to move the focus with `mod+hjkl`, which works great. Now, something I use quite a lot in i3 are tabbed workspaces. I do this for applications that I don't use simultaneously and I want to be maximized, like my email client and Slack. I don't think hyprland has tabbed containers, but it does have groups, which is similar enough.

However, something I haven't been able to, is having the same shortcut for navigate both groups and windows. In other words, if I move focus with `mod+hjkl` for tiled windows, I would like to use `mod+hl` to move focus inside a group as well. So far I haven't been able to achieve this. I have tried to map both group and window navigation to the same key:

```
# Move focus with mainMod + arrow keys
bind = $mainMod, h, changegroupactive, b
bind = $mainMod, h, movefocus, l
bind = $mainMod, l, changegroupactive, f
bind = $mainMod, l, movefocus, r
```

 but this leads to weird results when a group is placed next to a screen.

I wonder how others solving this use case: How do you achieve having multiple full-screen windows and navigating between them without having to use even more shortcuts?

**Comment by vaxerski** [comment]
you could script this

**Reply by nadiamoe** [reply]
Just to see if I understand your suggestion correctly, I can devise something like this:

1. Calling `hyperctl info activewindow` to check if the active window is a group (not sure if this info is returned)
2. If it is, calling `hyperctl dispatch changegroupactive`
3. If it is not, calling `hyperctl dispatch movefocus`

Is this similar to what you had in mind? Seems plausible but I'm just a tad worried about a possible race condition if the focus changes between the check and the dispatch commands. Probably it's not a big deal in practice though.

**Reply by thiagokokada** [reply]
Got it to work with this monstrosity:

```bash
activewindow="$(hyprctl activewindow -j)"
readonly activewindow
if ! jq -e '.grouped[]' <<< "$activewindow" >/dev/null; then
  hyprctl dispatch movefocus "$1"
elif [[ "$1" == l ]] && ${jq} -e '.address == .grouped[0]' <<< "$activewindow" >/dev/null; then
  hyprctl dispatch movefocus l
elif [[ "$1" == l ]]; then
  hyprctl dispatch changegroupactive b
elif [[ "$1" == r ]] && ${jq} -e '.address == .grouped[-1]' <<< "$activewindow" >/dev/null; then
  hyprctl dispatch movefocus r
else
  hyprctl dispatch changegroupactive f
fi
```

Warning: this is slow, adding ~50ms for each keystroke. Not the worst, but you can definitely few the impact once you change.

I definitely think this would be better written in a decent language like Go or Rust. But even better, it would be nice to have something integrated in Hyprland itself, like a `changegroupactiveormovefocus` dispatcher.

**Reply by thiagokokada** [reply]
For the mappings, I am only mapping left and right, something like:

```
bind=SUPER, left, exec, /nix/store/isha7ba1kz48caymp3dydnjsq3lw1q6x-changegroupactiveormovefocus l
bind=SUPER, right, exec, /nix/store/isha7ba1kz48caymp3dydnjsq3lw1q6x-changegroupactiveormovefocus r
bind=SUPER, H, exec, /nix/store/isha7ba1kz48caymp3dydnjsq3lw1q6x-changegroupactiveormovefocus l
bind=SUPER, L, exec, /nix/store/isha7ba1kz48caymp3dydnjsq3lw1q6x-changegroupactiveormovefocus r
```

**Reply by qboileau** [reply]
@thiagokokada Thank for the script, done a go implementation [here](https://github.com/qboileau/nix-config/tree/main/pkgs/hypr-i3-move)  that also add the move window inside/outside a group, (can extract in separate repository if needed)

**Comment by jthomaschewski** [comment]
@roobre have you found a solution in the meantime?

I have the exact same issue while trying out Hyprland as along time i3/sway user.  
Will try do write a small script for this task but it feels like this would be a better fit for a built-in feature.

Alongside navigation it would be great to also be able to move windows INTO a group using the same shortcut as moving windows around directional.

e.g.
```
bind = SUPER_SHIFT, RIGHT, moveintogroup, r
bind = SUPER_SHIFT, RIGHT, movewindow, r
```

**Reply by nadiamoe** [reply]
I _think_ the proposed script approach, i.e. binding a custom script to the window movement letters might work, but I haven't tried this out yet as I try to figure out another rock: I like how in i3 my desktops are tabbed by default, and I switch to tiled when I need it (e.g. terminal, specific operations). I haven't jumped into hyprland yet as I'm not sure how to work around that second problem :(

**Reply by rawat-sahil** [reply]
I am also trying hyprland for the first time and have been i3 user for a long time. I have been liking the hyprland so far but this thing is making me go nuts. Tabs were a major part of my work flow where i would just maximise the screen that i need and split them whenever i needed both windows. Have you guys been able to figure it out so far?

**Comment by 3333juul** [comment]
I have a script for this purpose. It remaps the keys if only grouped windows are in the active workspace, and reverts them back to the previous ones if not. I use cyclenext/cycleprev and swapnext/swapprev dispatchers so you need to adjust the script to yourself if you use different dispatchers for navigation.

```
exec-once = ~/.config/hypr/scripts/groupbind
```

```
#!/bin/sh

last_commands=""

handle() {
  case $1 in
    activewindow*)
        # Fetching data from hyprctl
        monitors=$(hyprctl -j monitors)
        clients=$(hyprctl -j clients)

        # Searching for the active monitor
        active_monitor=$(echo "$monitors" | jq '.[] | select(.focused==true)')

        # Checking if there is a specialWorkspace
        if [ -n "$(echo "$active_monitor" | jq -r '.specialWorkspace.name')" ]; then
            active_workspace=$(echo "$active_monitor" | jq -r '.specialWorkspace.name')
        else
            active_workspace=$(echo "$active_monitor" | jq -r '.activeWorkspace.name')
        fi

        # Checking if windows in the active workspace have 'grouped'
        windows_in_active_workspace=$(echo "$clients" | jq -c ".[] | select(.workspace.name==\"$active_workspace\")")
        grouped_windows_in_active_workspace=$(echo "$windows_in_active_workspace" | jq -c 'select(.grouped | length > 0)')

        # Checking if windows in the group are the only ones in the active workspace
        if [ "$(echo "$grouped_windows_in_active_workspace" | jq -s '. | length')" -gt 0 ] && [ "$(echo "$windows_in_active_workspace" | jq -s '. | length')" == "$(echo "$grouped_windows_in_active_workspace" | jq -s '. | length')" ]; then
            new_commands="group"
            if [ "$last_commands" != "$new_commands" ]; then
                hyprctl keyword unbind SUPER,J
                hyprctl keyword unbind SUPER,K
                hyprctl keyword unbind SUPER SHIFT,J
                hyprctl keyword unbind SUPER SHIFT,K
                hyprctl keyword bind SUPER,J, changegroupactive, f
                hyprctl keyword bind SUPER,K, changegroupactive, b
                hyprctl keyword bind SUPER SHIFT,J, movegroupwindow, f
                hyprctl keyword bind SUPER SHIFT,K, movegroupwindow, b
                last_commands=$new_commands
            fi
        else
            new_commands="layout"
            if [ "$last_commands" != "$new_commands" ]; then
                hyprctl keyword unbind SUPER,J
                hyprctl keyword unbind SUPER,K
                hyprctl keyword unbind SUPER SHIFT,J
                hyprctl keyword unbind SUPER SHIFT,K
                hyprctl keyword bind SUPER,J,layoutmsg,cyclenext
                hyprctl keyword bind SUPER,K,layoutmsg,cycleprev
                hyprctl keyword bind SUPER SHIFT,J,layoutmsg,swapnext
                hyprctl keyword bind SUPER SHIFT,K,layoutmsg,swapprev
                last_commands=$new_commands
            fi
        fi
    ;;
  esac
}

socat -U - UNIX-CONNECT:/tmp/hypr/$HYPRLAND_INSTANCE_SIGNATURE/.socket2.sock | while read -r line; do handle "$line"; done
```

**Reply by C0DK** [reply]
Maybe hyprland has changed, or i am doing something wrong, but i get an error when running it manually:
```
2025/03/26 08:14:02 socat[20960] E connect(, AF=1 "/tmp/hypr/12f9a0d0b93f691d4d9923716557154d74777b0a_1742971409_881083663/.socket2.sock", 87): No such file or directory
```

exec from the terminal just gives an `ok`:
```
➜ hyprctl dispatch exec ~/.config/hypr/scripts/groupbind
ok
```

It seems like there is no `/tmp/hypr/` folder at all_ is this nix related maybe possibly?

**Reply by fufexan** [reply]
Try `$XDG_RUNTIME_DIR/hypr/$HYPRLAND_INSTANCE_SIGNATURE` instead.

**Comment by duckie** [comment]
It would be so great to have it integrated. I really do not like the script approach, this feels clumsy for something so simple.

**Reply by amitbha** [reply]
Try sway.

**Comment by user451421541757324** [comment]
Guys, now that we have Lua, I tried to put something together:
```
hl.config(
    {
        dwindle = {
            preserve_split = true,
            force_split = 2,
            smart_split = false,
            split_width_multiplier = 1.0,
            use_active_for_splits = true
        },
        group = {
            insert_after_current = false
        }
    }
)

-- toggle group
hl.bind("SUPER + W", hl.dsp.group.toggle())

-- change focus and move windows
local directions = {"left", "right", "up", "down"}
local group_nav = {left = hl.dsp.group.prev, right = hl.dsp.group.next}

for _, dir in ipairs(directions) do
    local short_dir = dir:sub(1, 1)
    local nav = group_nav[dir]

    -- focus
    hl.bind(
        "SUPER + " .. dir,
        function()
            local win = hl.get_active_window()
            local grp = win and win.group

            if grp then
                if (grp.current_index == 1 and short_dir == "l") then
                    hl.dispatch(hl.dsp.focus({direction = short_dir}))
                elseif (grp.current_index == grp.size and short_dir == "r") then
                    hl.dispatch(hl.dsp.focus({direction = short_dir}))
                else
                    local next_index
                    if short_dir == "l" then
                        next_index = grp.current_index - 1
                    else
                        next_index = grp.current_index + 1
                    end

                    hl.dispatch(hl.dsp.group.active({index = next_index}))
                end
            else
                hl.dispatch(hl.dsp.focus({direction = short_dir}))
            end
        end
    )

    -- move
    hl.bind(
        "SUPER + SHIFT + " .. dir,
        function()
            hl.config({dwindle = {smart_split = true}})

            local win = hl.get_active_window()
            local grp = win and win.group

            local move_window_normally =
                not grp or short_dir == "u" or short_dir == "d" or (short_dir == "l" and grp.current_index == 1) or
                (short_dir == "r" and grp.current_index == grp.size)

            if move_window_normally then
                hl.dispatch(hl.dsp.window.move({direction = short_dir, group_aware = true}))
            else
                hl.dispatch(hl.dsp.group.move_window({forward = dir == "right"}))
            end

            hl.config({dwindle = {smart_split = false}})
        end
    )
end
```

Unfortunately, it doesn't work very well:

- windows are always inserted at the end of the group (or after the last active group window if `insert_after_current` is commented out)
- moving the focus inside a group always targets the last active window

If anyone smarter than me wants to give it a try...

You can find the Hyprland Lua documentation in `/usr/share/hypr/stubs/hl.meta.lua`.
